src/compute/src/render/continual_task.rs

```diff
@@ -473,59 +472,160 @@ fn continual_task_sink<G: Scope<Timestamp = Timestamp>>(
             }
         }
 
-        let to_insert_input = to_append_input.map(OpEvent::ToAppend);
+        let to_append_input = to_append_input.map(OpEvent::ToAppend);
         let append_times_input = append_times_input.map(OpEvent::AppendTimes);
-        let mut op_inputs = futures::stream::select(to_insert_input, append_times_input);
+        let mut op_events = futures::stream::select(to_append_input, append_times_input);
 
         let mut state = SinkState::new();
+        state.output_progress = write_handle.shared_upper();
         loop {
-            if PartialOrder::less_than(&*output_frontier.borrow(), &state.output_progress) {
-                output_frontier.borrow_mut().clear();
-                output_frontier
-                    .borrow_mut()
-                    .extend(state.output_progress.iter().cloned());
-            }
-            let Some(event) = op_inputs.next().await else {
+            let Some(event) = op_events.next().await else {
                 // Inputs exhausted, shutting down.
-                output_frontier.borrow_mut().clear();
                 return;
             };
             event.apply(&mut state);
             // Also drain any other events that may be ready.
-            while let Some(Some(event)) = op_inputs.next().now_or_never() {
+            while let Some(Some(event)) = op_events.next().now_or_never() {
                 event.apply(&mut state);
             }
             debug!("ct_sink about to process {:?}", state);
-            let Some((new_upper, to_append)) = state.process() else {
+            let Some((new_upper, mut to_append)) = state.process() else {
                 continue;
             };
             debug!("ct_sink got write {:?}: {:?}", new_upper, to_append);
+            output.give_container(&cap, &mut to_append);
+            if let Some(new_upper) = new_upper.as_option() {
+                cap.downgrade(new_upper);
+            };
```

Comment:
Why don't we need to drop the capability when `new_upper` is the empty frontier? Are we guaranteed to take the "Inputs exhausted, shutting down" branch next in that case?